File: sort.js
var _data = [99,5,2,1,31,15,25,12,35,33,56,78,21,34,30,55];

/* ---Bubble Sort---*/
function bubbleSort(input){
  input = input.slice(0);
  for (var i=0;i<input.length;i++){
    for (var j=0;j<input.length-i-1;j++){
      if (input[j]>input[j+1]){
        var temp = input[j];
        input[j] = input[j+1];
        input[j+1] = temp;
      }
    }
  }
  return input;
}

/* ---Insertion Sort, efficent for small array---*/
function insertSort(input){
  input = input.slice(0);
  for (var i=1;i<input.length;i++){
    var temp = input[i];
    var j = i - 1;
    while(j>=0 && input[j]>temp){
      input[j+1] = input[j];
      j--;
    }
    input[j+1] = temp;
  }
  return input;
}

/* ---Merge Sort---*/
function _merge(a, start, middle, end, store){
  var ia = start;
  var ib = middle;
  for(var i=start;i<end;i++){
    if ((ia<middle && ib<end && a[ia]<a[ib]) || (ia<middle && ib>=end)){
      store[i] = a[ia];
      ia++;
    }else{
      store[i] = a[ib];
      ib++;
    }
  }
}

function _splitmerge(a, start, end, store){
  if (end - start < 2) return;
  var middle = Math.floor((start + end) / 2);
  _splitmerge(a, start, middle, store);
  _splitmerge(a, middle, end, store);
  _merge(a, start, middle, end, store);
  // copy back
  for(var i=start;i<end;i++){
    a[i] = store[i];
  }
}

function mergeSortTopDown(input){
  // recusion version of merge sort (top-down)
  // although in theory, merge sort is O(logn) algorithm
  // but function recusion in js will cost too much other spending
  // try bottom-up non recusion version
  input = input.slice(0);
  var store = new Array(input.length);
  _splitmerge(input, 0, input.length, store);
  return input;
}

function mergeSortBottomUp(input){
  input = input.slice(0);
  var bak = new Array(input.length);
  var n = input.length;
  var handle = input;
  var store = bak;
  for (var width=1; width<n; width=2*width){
    for (var i=0; i<n; i=i+2*width){
      _merge(handle, i, Math.min(i+width,n), Math.min(i+2*width,n), store);
    }
    // swith for next run
    if (handle===input) handle = bak; else if (handle===bak) handle = input;
    if (store===input) store = bak; else if (store===bak) store = input;
  }
  return input;
}

/* ---Quick Sort, Simple version: pivot not the median value, don't consider equal pivot condition---*/
function _qsort(a, left, right){
  if (left < right){
    var pivotIndex = _partition(a, left, right);
    _qsort(a, left, pivotIndex - 1);
    _qsort(a, pivotIndex + 1, right);
  }
}

function _partition(a, left, right){
  var pivotIndex = right; // not optimize for pivot selection
  var k = left;
  var tmp;
  for (var i=left;i<=right-1;i++){
    if (a[i]<=a[pivotIndex]){
      tmp = a[i];
      a[i] = a[k];
      a[k] = tmp;
      k++;
    }
  }
  tmp = a[k];
  a[k] = a[pivotIndex];
  a[pivotIndex] = tmp;
  return k;
}

function quickSort(input){
  input = input.slice(0);
  _qsort(input, 0, input.length-1);
  return input;
}

/* ---Bucket Sort---*/
function bucketSort(input){
  input = input.slice(0);
  return input;
}

console.log("bubble sort");
var begin = new Date().getTime();
console.log(bubbleSort(_data).join());
for(var i=0;i<1000000;i++) bubbleSort(_data);
console.log("cost:" + (new Date().getTime() - begin).toString());

process.stdout.write('\n');
console.log("insertion sort");
var begin = new Date().getTime();
console.log(insertSort(_data).join());
for(var i=0;i<1000000;i++) insertSort(_data);
console.log("cost:" + (new Date().getTime() - begin).toString());

process.stdout.write('\n');
console.log("merge sort");
var begin = new Date().getTime();
console.log(mergeSortBottomUp(_data).join());
for(var i=0;i<1000000;i++) mergeSortBottomUp(_data);
console.log("cost:" + (new Date().getTime() - begin).toString());

process.stdout.write('\n');
console.log("quick sort");
var begin = new Date().getTime();
console.log(quickSort(_data).join());
for(var i=0;i<1000000;i++) quickSort(_data);
console.log("cost:" + (new Date().getTime() - begin).toString());

process.stdout.write('\n');
console.log("bucket sort");
var begin = new Date().getTime();
console.log(bucketSort(_data).join());
for(var i=0;i<1000000;i++) bucketSort(_data);
console.log("cost:" + (new Date().getTime() - begin).toString());

File: sort.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

#define MIN(X,Y) ((X) < (Y) ? (X) : (Y))

void printArray(int *a, int n){
  int i=0;
  for (i=0;i<n;i++){
    printf("%d, ", a[i]);
  }
  printf("\n");
}

void duplicateArray(int *src, int *dest, int n){
  int i=0;
  for(i=0;i<n;i++){
    dest[i] = src[i];
  }
}

/* ---Bubble Sort---*/
void bubbleSort(int *input, int n){
  int i,j;
  for(i=0;i<n;i++){
    for(j=0;j<n-i-1;j++){
      if(input[j]>input[j+1]){
        int tmp = input[j];
        input[j] = input[j+1];
        input[j+1] = tmp;
      }
    }
  }
}

/* ---Insertion Sort, efficent for small array---*/
void insertSort(int *input, int n){
  int i,j;
  for(i=1;i<n;i++){
    int tmp = input[i];
    j = i - 1;
    while(j>=0 && input[j]>tmp){
      input[j+1] = input[j];
      j--;
    }
    input[j+1] = tmp;
  }
}

/* ---Merge Sort, bottom up---*/
void _merge(int *a, int start, int middle, int end, int *store){
  int ia = start;
  int ib = middle;
  int i;
  for (i=start;i<end;i++){
    if ((ia<middle && ib<end && a[ia]<a[ib]) || (ia<middle && ib>=end)){
      store[i] = a[ia];
      ia++;
    }else{
      store[i] = a[ib];
      ib++;
    }
  }
}

void mergeSort(int *input, int n){
  int *bak = malloc(sizeof(int)*n);
  int *handle = input;
  int *store = bak;
  int width,i;
  for(width=1;width<n;width=2*width){
    for(i=0;i<n;i=i+2*width){
      _merge(handle, i, MIN(i+width,n), MIN(i+2*width,n),store);
    }
    if (handle==input) handle = bak; else if (handle==bak) handle = input;
    if (store==input) store = bak; else if (store==bak) store = input;
  }
  free(bak);
}

/* ---Quick Sort---*/
int _partition(int *a, int left, int right){
  int pivotIndex = right;
  int k = left;
  int tmp,i;
  for(i=left;i<=right-1;i++){
    if (a[i]<=a[pivotIndex]){
      tmp = a[i];
      a[i] = a[k];
      a[k] = tmp;
      k++;
    }
  }
  tmp = a[k];
  a[k] = a[pivotIndex];
  a[pivotIndex] = tmp;
  return k;
}

void _qsort(int *a, int left, int right){
  if (left<right){
    int pivotIndex = _partition(a, left, right);
    _qsort(a, left, pivotIndex-1);
    _qsort(a, pivotIndex+1 ,right);
  }
}

void quickSort(int *a, int n){
  _qsort(a, 0, n-1);
}

/* ---Radix Sort---*/
void radixSort(int *a, int n){
  int k = 0;
  int i;
  for(i=0;i<n;i++){
    if (a[i]>k) k = a[i];
  }
  int *store = malloc(sizeof(int)*n);
  int e = 1;
  int *workArray = a;
  int *tmpArray = store;
  int zc = 0;
  while(k/e>0){
    int buckets[10] = {0};
    for(i=0;i<n;i++){
      buckets[(workArray[i]/e)%10]++;
    }
    for(i=1;i<10;i++){
      buckets[i] += buckets[i-1];
    }
    for(i=n-1;i>=0;i--){
      tmpArray[--buckets[(workArray[i]/e)%10]] = workArray[i];
    }
    if(workArray==a) workArray = store; else if(workArray==store) workArray = a;
    if(tmpArray==a) tmpArray = store; else if(tmpArray==store) tmpArray = a;
    e*=10;
    zc++;
  }
  // copy back
  if(zc%2>0){
    for(i=0;i<n;i++){
      a[i] = store[i];
    }
  }
  free(store);
}

/* --- Begin Test---*/
void test(void(*fun)(int*, int), int *input, int n, char *text){
  printf("%s\n", text);
  // output result
  int *array = malloc(sizeof(int)*n);
  duplicateArray(input, array, n);
  fun(array, n);
  printArray(array, n);
  free(array);
  // test performance
  int *ta = malloc(sizeof(int)*n);
  clock_t start = clock();
  int i=0;
  for(i=0;i<100000;i++){
    duplicateArray(input, ta, n);
    fun(ta, n);
  }
  clock_t end = clock();
  printf("Cost: %f\n", (double)(end-start)/CLOCKS_PER_SEC);
  free(ta);
  printf("\n");
}

int main(){
  int n=200;
  int testData[200] = {0};
  int i = 0;
  srand(time(NULL));
  for(i=0;i<n;i++){
    testData[i] = rand()%200 + 1;
  }
  printf("Ori Array:\n");
  printArray(&testData[0], n);
  printf("\n");
  test(&bubbleSort, &testData[0], n, "Bubble Sort");
  test(&insertSort, &testData[0], n, "Insertion Sort");
  test(&mergeSort, &testData[0], n, "Merge Sort");
  test(&quickSort, &testData[0], n, "Quick Sort");
  test(&radixSort, &testData[0], n, "Radix Sort");
  return 0;
}
